File: backend/app/geo.py
"""Straight-line distances (docs/06 §4), shared by the API and the seed sync."""

from math import asin, cos, radians, sin, sqrt

EARTH_RADIUS_KM = 6371


def haversine_km(lat1: float, lon1: float, lat2: float, lon2: float) -> int:
    """Great-circle distance between two points, rounded to whole kilometres."""
    d_lat = radians(lat2 - lat1)
    d_lon = radians(lon2 - lon1)
    h = sin(d_lat / 2) ** 2 + cos(radians(lat1)) * cos(radians(lat2)) * sin(d_lon / 2) ** 2
    return round(2 * EARTH_RADIUS_KM * asin(sqrt(h)))

File: frontend/src/lib/geo.ts
/** Straight-line distances between area centres (docs/06 §4): haversine, Earth radius 6371 km. */

export interface LatLon {
  readonly lat: number
  readonly lon: number
}

const EARTH_RADIUS_KM = 6371
const rad = (deg: number) => (deg * Math.PI) / 180

export function distanceKm(a: LatLon, b: LatLon): number {
  const dLat = rad(b.lat - a.lat)
  const dLon = rad(b.lon - a.lon)
  const h =
    Math.sin(dLat / 2) ** 2 + Math.cos(rad(a.lat)) * Math.cos(rad(b.lat)) * Math.sin(dLon / 2) ** 2
  return Math.round(2 * EARTH_RADIUS_KM * Math.asin(Math.sqrt(h)))
}

/** A copy ordered from the nearest to the farthest from `from` (stable for ties). */
export function byDistance<T extends LatLon>(places: readonly T[], from: LatLon): T[] {
  return places
    .map((place, index) => ({ place, index, km: distanceKm(from, place) }))
    .sort((x, y) => x.km - y.km || x.index - y.index)
    .map((x) => x.place)
}
